File: app/src/main/java/org/mesonet/app/baseclasses/RecyclerViewAdapter.java
package org.mesonet.app.baseclasses;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;



public abstract class RecyclerViewAdapter<TData, TRecyclerViewHolder extends RecyclerViewHolder> extends RecyclerView.Adapter<TRecyclerViewHolder>
{
    protected boolean mUsePagination = false;
    protected List<TData> mDataItems = new ArrayList<>();
    protected RecyclerView mRecyclerView;
    protected List<TRecyclerViewHolder> mRecyclerViewHolders = new ArrayList<>();

    protected List<ScrolledToEndListener> mScrolledToEndListeners = new ArrayList<>();

    private RecyclerView.OnScrollListener kOnScrollListener;



    public RecyclerViewAdapter()
    {
        setHasStableIds(true);

        kOnScrollListener = new RecyclerView.OnScrollListener()
        {
            @Override
            public void onScrolled(RecyclerView inRecyclerView, int inDx, int inDy)
            {
                super.onScrolled(inRecyclerView, inDx, inDy);

                RecyclerView.LayoutManager layoutManager = mRecyclerView.getLayoutManager();

                if (mDataItems != null)
                {
                    if (((LinearLayoutManager) layoutManager).findLastCompletelyVisibleItemPosition() == mDataItems.size())
                    {
                        for (ScrolledToEndListener scrolledToEndListener : mScrolledToEndListeners)
                            scrolledToEndListener.OnScrolledToEnd();
                    }
                }
            }
        };
    }



    @Override
    public int getItemCount()
    {
        int result = 0;

        if (mDataItems != null)
            result = mDataItems.size();

        if (mUsePagination)
            result++;

        return result;
    }



    @Override
    public void onAttachedToRecyclerView(RecyclerView inRecyclerView)
    {
        mRecyclerView = inRecyclerView;
        mRecyclerView.addOnScrollListener(kOnScrollListener);
    }



    @Override
    public void onBindViewHolder(TRecyclerViewHolder inHolder, int inPosition)
    {
        if (inPosition < mDataItems.size())
        {
            TData data = mDataItems.get(inPosition);

            inHolder.SetData(data);
        }
    }



    @Override
    public long getItemId(int inPostition)
    {
        return inPostition;
    }



    public void SetItems(final List<TData> inDataItems)
    {
        SetItems(inDataItems, 0);
    }



    public void SetItems(List<TData> inDataItems, int inTotalItems)
    {
        if (mDataItems == inDataItems)
        {
            return;
        }

        mDataItems = inDataItems;
        mUsePagination = (mDataItems == null ? 0 : mDataItems.size()) < inTotalItems;
        notifyDataSetChanged();
    }



    public void AddItems(List<TData> inDataItems)
    {
        AddItems(inDataItems, 0);
    }

    public void AddItems(List<TData> inDataItems, int inTotalItems)
    {
        if (mDataItems == null)
            mDataItems = new ArrayList<>();

        mDataItems.addAll(inDataItems);
        mUsePagination = mDataItems.size() < inTotalItems;
        notifyDataSetChanged();
    }

    public void ClearItems()
    {
        if (mDataItems != null)
        {
            mDataItems.clear();
            mUsePagination = false;
            notifyDataSetChanged();
        }
    }



    public int GetItemCount()
    {
        int count = getItemCount();

        if (mUsePagination)
            count--;

        return count;
    }



    public void AddScrolledToEndListener(final ScrolledToEndListener inScrolledToEndListener)
    {
        mScrolledToEndListeners.add(inScrolledToEndListener);
    }


    public void RemoveScrolledToEndListener(final ScrolledToEndListener inScrolledToEndListener){
        mScrolledToEndListeners.remove(inScrolledToEndListener);
    }


    public void AddViewHolder(TRecyclerViewHolder inTRecyclerViewHolder)
    {
        mRecyclerViewHolders.add(inTRecyclerViewHolder);
    }

    public List<TRecyclerViewHolder> GetViewHolders()
    {
        return mRecyclerViewHolders;
    }



    @Override
    public void finalize()
    {
        mRecyclerView.removeOnScrollListener(kOnScrollListener);

        kOnScrollListener = null;
        mScrolledToEndListeners = null;
        mRecyclerView = null;
    }



    public interface ScrolledToEndListener
    {
        void OnScrolledToEnd();
    }
}

File: app/src/main/java/org/mesonet/app/baseclasses/RecyclerViewAdapter.kt
package org.mesonet.app.baseclasses

import android.databinding.ViewDataBinding
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

import java.util.ArrayList
import android.content.ContextWrapper
import android.app.Activity




abstract class RecyclerViewAdapter<TData, TRecyclerViewHolder : RecyclerViewHolder<TData, ViewDataBinding>> : RecyclerView.Adapter<TRecyclerViewHolder>() {
    protected var mUsePagination = false
    protected var mDataItems: MutableList<TData>? = ArrayList()
    protected var mRecyclerView: RecyclerView? = null
    protected var mRecyclerViewHolders: MutableList<TRecyclerViewHolder> = ArrayList()

    protected var mScrolledToEndListeners: MutableList<ScrolledToEndListener>? = ArrayList()

    private var kOnScrollListener: RecyclerView.OnScrollListener? = null

    init {
        setHasStableIds(true)

        kOnScrollListener = object : RecyclerView.OnScrollListener() {
            override fun onScrolled(inRecyclerView: RecyclerView, inDx: Int, inDy: Int) {
                super.onScrolled(inRecyclerView, inDx, inDy)

                val layoutManager = mRecyclerView?.layoutManager

                if (mDataItems != null) {
                    if ((layoutManager as LinearLayoutManager).findLastCompletelyVisibleItemPosition() == mDataItems?.size) {
                        for (scrolledToEndListener in mScrolledToEndListeners?: ArrayList())
                            scrolledToEndListener.OnScrolledToEnd()
                    }
                }
            }
        }
    }


    override fun getItemCount(): Int {
        var result = 0

        if (mDataItems != null)
            result = mDataItems?.size ?: 0

        if (mUsePagination)
            result++

        return result
    }


    override fun onAttachedToRecyclerView(inRecyclerView: RecyclerView) {
        mRecyclerView = inRecyclerView
        if(kOnScrollListener != null)
            mRecyclerView?.addOnScrollListener(kOnScrollListener ?: object: RecyclerView.OnScrollListener(){})
    }


    override fun onBindViewHolder(inHolder: TRecyclerViewHolder, inPosition: Int) {
        if (inPosition < mDataItems?.size ?: 0) {
            val data = mDataItems?.get(inPosition)

            inHolder.SetData(data)
        }
    }


    override fun getItemId(inPostition: Int): Long
    {
        return inPostition.toLong()
    }


    internal fun SetItems(inDataItems: MutableList<TData>?, inTotalItems: Int = 0)
    {
        if (mDataItems === inDataItems)
            return

        mDataItems = inDataItems
        mUsePagination = (mDataItems?.size ?: 0) < inTotalItems

        dataSetChanged()
    }


    internal fun AddItems(inDataItems: List<TData>, inTotalItems: Int = 0) {
        if (mDataItems == null)
            mDataItems = ArrayList()

        mDataItems?.addAll(inDataItems)
        mUsePagination = mDataItems?.size ?: 0 < inTotalItems

        dataSetChanged()
    }


    internal fun ClearItems() {
        mDataItems?.clear()
        mUsePagination = false

        dataSetChanged()
    }

    private fun dataSetChanged() {
        GetActivity()?.runOnUiThread {
            notifyDataSetChanged()
        }
    }


    private fun GetActivity(): Activity? {
        var context = mRecyclerView?.context
        while (context is ContextWrapper) {
            if (context is Activity) {
                return context
            }
            context = context.baseContext
        }
        return null
    }


    internal fun GetItemCount(): Int {
        var count = itemCount

        if (mUsePagination)
            count--

        return count
    }


    internal fun AddScrolledToEndListener(inScrolledToEndListener: ScrolledToEndListener) {
        mScrolledToEndListeners?.add(inScrolledToEndListener)
    }


    internal fun RemoveScrolledToEndListener(inScrolledToEndListener: ScrolledToEndListener) {
        mScrolledToEndListeners?.remove(inScrolledToEndListener)
    }


    internal fun AddViewHolder(inTRecyclerViewHolder: TRecyclerViewHolder) {
        mRecyclerViewHolders.add(inTRecyclerViewHolder)
    }

    internal fun GetViewHolders(): List<TRecyclerViewHolder> {
        return mRecyclerViewHolders
    }


    fun finalize() {
        mRecyclerView?.removeOnScrollListener(kOnScrollListener ?: object: RecyclerView.OnScrollListener(){})

        kOnScrollListener = null
        mScrolledToEndListeners = null
        mRecyclerView = null
    }


    interface ScrolledToEndListener {
        fun OnScrolledToEnd()
    }
}